# **MÓDULO 15 - Exercício**
# Análise  - A segunda etapa da Pré Modelagem

# 1) O primeiro exercício é o de salvar a base que criaram na atividade do módulo anterior em csv e abrir ela neste arquivo.
Igual fizemos no início do módulo atual no início da primeira aula.

In [2]:
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import plotly.express as px

UNIVARIADA

* Genero - Ok
* Idoso - OK
* Tempo como cliente -> Histograma
* Churn -> Grafico de rosca -> Ok
* Tipo Contrato -> Grafico de barras -> Ok

-----

BIVARIADA

- Pagamento_Mensal e Churn
- Casado, dependentes com Churn ---> Mapa de arvóre
- Idoso e Churn
- Tipo_Contrato e Churn
- Genero e Churn
- Tempo como cliente e Churn
- Serviço Internet, Servico_Seguranca, Suporte Tecnico Streaming TV e Churn ---> Mapa de arvóre



In [3]:
#Seu código de importação aqui
df = pd.read_csv('Dados/CHURN_TELECON_TAREFA_TRATADO_CONTINUACAO-1.csv', delimiter = ',')
df.head()

Genero  Idoso Casado Dependentes  Tempo_como_Cliente Servico_Internet  \
0  Female      0    Yes          No                   1              DSL   
1    Male      0     No          No                  34              DSL   
2    Male      0     No          No                   2              DSL   
3    Male      0     No          No                  45              DSL   
4  Female      0     No          No                  52               NO   

     Servico_Seguranca      Suporte_Tecnico          StreamingTV  \
0                   No                   No                   No   
1                  Yes                   No                   No   
2                  Yes                   No                   No   
3                  Yes                  Yes                   No   
4  No internet service  No internet service  No internet service   

    Tipo_Contrato           Metodo_Pagamento  Pagamento_Mensal  Total_Pago  \
0  Month-to-month           Electronic check             29.85       29.85   
1        One year               Mailed check             56.95     1889.50   
2  Month-to-month               Mailed check             53.85      108.15   
3        One year  Bank transfer (automatic)             71.45     1840.75   
4        One year               Mailed check             71.45     1022.95   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4    No

# 2) Comece pela análise univariada:

A) Utilize a função describe no seu dataframe, veja os insights que consegue retirar.

B) Já é possível identificar variáveis com possíveis outliers? Se sim, quais?

C) Plot gráficos que considerar importante para completar sua análise univariada. (Lembrando que sua variável preditora é o churn). Não se esqueça de trazer insights de cada gráfico plotado. Utilize pelo menos 4 variáveis distintas.

D) Verifique se os dados das variáveis Booleanas são balanceados ou não.

## Exercício A

In [3]:
#Seu código aqui
df.describe()

Idoso  Tempo_como_Cliente  Pagamento_Mensal   Total_Pago
count  2488.000000         2488.000000       2488.000000  2488.000000
mean      0.161576           32.352090         66.342564  2290.087178
std       0.368135           24.636885         28.051718  2264.585643
min       0.000000            0.000000         18.400000    18.800000
25%       0.000000            8.000000         45.400000   402.337500
50%       0.000000           29.000000         71.450000  1402.900000
75%       0.000000           56.000000         87.712500  3867.162500
max       1.000000           72.000000        118.650000  8564.750000

#### insights a partir da função describe():
- Mediana do tempo de clientes conosco: 29 meses
- Média de mensalidade contratada: 66,34
- Mediana total pago por cliente: 1402,90
- Mensalidade mais baixa contratada: 18,40
- Mensalidade mais cara contratado: 118,65
- Total de clientes: 2488

Desenvolvendo...

A base analisada é composta por **2488 clientes.**
O **tempo típico de permanência** é de **29 meses**, sendo o **ticket médio** de mensalidades em torno de **66,34**.
O **total acumulado pago por cliente** é, em termos centrais, de **1402,90.**
**As mensalidades** variam entre **18,40 e 118,65**

## Exercício B

Uma coluna que possivelmente teria outliers seria a Total_Pago, mas não é possível ter certeza apenas pela função describe()

As demais colunas, acredito que estão de acordo

## Exercício C

### Coluna Gênero

In [182]:
contagem_genero = df['Genero'].value_counts().reset_index(name = 'Contagem')
contagem_genero['Percentual'] = ((contagem_genero['Contagem'] / len(df)) * 100).round(2)

print('Valores Brutos: \n', contagem_genero, '\n')
fig = px.bar(
    contagem_genero,
    x = 'Genero',
    y = 'Percentual',
    title = 'Gêneros de Clientes [%]',
    template = 'plotly',
    width = 1000,
    height = 600,
    labels = {'Genero': 'Gênero'},
    color = 'Genero',
    color_discrete_map = {
        'Female': '#FD8AFF',
        'Male': '#6978FF'
    },
    text_auto = True
)

fig.update_layout(
    font = dict(
        size = 16
    ),
    hoverlabel = dict(
        font = dict(
            size = 14
        )
    )
)

fig.update_traces(
    textfont_size = 18,
    textposition = 'outside',
)

fig.show()

Valores Brutos: 
    Genero  Contagem  Percentual
0    Male      1269        51.0
1  Female      1219        49.0 



### **Insight**

Podemos observar que os gêneros dos clientes é bem balanceado, sendo **51% homens e 49% mulheres**

### Coluna Idoso

In [219]:
# CRIANDO UM DATAFRAME PARALELO ---> MELHORAR A VISUALIZAÇÃO
df_idoso_tratado = df.copy()
df_idoso_tratado['Idoso'] = df_idoso_tratado['Idoso'].replace({0: 'Não Idoso', 1: 'Idoso'})

# CALCULANDO OS VALORES
contagem_idosos = df_idoso_tratado['Idoso'].value_counts().reset_index(name = 'Contagem')
contagem_idosos['Percentual'] = ((contagem_idosos['Contagem'] / len(df)) * 100).round(2)

print(f'Valores Brutos: \n {contagem_idosos}\n')

# PLOTANDO O GRÁFICO
fig = px.pie(
    contagem_idosos,
    values = 'Percentual',
    names = 'Idoso',
    template = 'plotly',
    hole = .5,
    title = 'Clientes Idosos [%]',
    width = 1000,
    height = 600,
    category_orders = {
        'Idoso': ['Idoso', 'Não Idoso']
    },
    color = 'Idoso',
    color_discrete_map = {
       'Idoso': '#6978FF',
       'Não Idoso': '#7f7f7f'
    }
)

fig.update_layout(
    font = dict(
        size = 20
    ),
    legend = dict( # ALTERANDO A LEGENDA DO GRÁFICO
    x = 0.8, # POSIÇÃO HORIZONTAL
    y = 1, # POSIÇÃO VERTICAL
    ),
    hoverlabel = dict(
        font = dict(
            color = 'white',
            size = 16
        )
    )
)

fig.update_traces(
    textfont = dict(
        color = 'white',
    )
)

fig.show()

Valores Brutos: 
        Idoso  Contagem  Percentual
0  Não Idoso      2086       83.84
1      Idoso       402       16.16



### **Insight**

Podemos observar que 83% do clientes não são pessoas idosas

### Coluna Tempo_Como_Cliente

In [185]:
df_tempo = df.copy()

valor_maximo = df_tempo['Tempo_como_Cliente'].max()
bins = list(range(0, valor_maximo + 7, 6))
labels = [f'{b} a {b + 5}' for b in bins[:-1]]

df_tempo['Faixa'] = pd.cut(
    df_tempo['Tempo_como_Cliente'],
    bins = bins,
    labels = labels,
    right = False,  # <--- CORREÇÃO: Mudar de True para False
    include_lowest = True
)

tabela_faixas = df_tempo['Faixa'].value_counts().sort_index().reset_index(name = 'Contagem')
tabela_faixas['Percentual'] = ((tabela_faixas['Contagem'] / len(df_tempo)) * 100).round(2)
print(tabela_faixas, '\n')

fig = px.bar(
    tabela_faixas,
    x = 'Faixa',
    y = 'Percentual',
    template = 'plotly',
    title = 'Distribuição de Clientes [%] Por Faixa de Tempo',
    width = 1000,
    height = 600
)

fig.update_layout(
    bargap = 0.05,
    font = dict(
        size = 14
    )
)

fig.show()

      Faixa  Contagem  Percentual
0     0 a 5       495       19.90
1    6 a 11       240        9.65
2   12 a 17       209        8.40
3   18 a 23       165        6.63
4   24 a 29       148        5.95
5   30 a 35       163        6.55
6   36 a 41       115        4.62
7   42 a 47       147        5.91
8   48 a 53       146        5.87
9   54 a 59       128        5.14
10  60 a 65       162        6.51
11  66 a 71       232        9.32
12  72 a 77       138        5.55 



### **Insight**

Podemos observar que cerca de 20% dos clientes se encontram na faixa de 0 a 5 meses com um pequeno destaque de 9% dos nossos clientes em 66 a 71 meses.

### Coluna Tipo Contrato

In [222]:
contagem_tipo_contrato = df['Tipo_Contrato'].value_counts().reset_index(name = 'Contagem')
contagem_tipo_contrato['Percentual'] = ((contagem_tipo_contrato['Contagem'] / len(df)) * 100).round(2)
print(contagem_tipo_contrato, '\n')

fig = px.bar(
    contagem_tipo_contrato,
    x = 'Tipo_Contrato',
    y = 'Percentual',
    template = 'plotly',
    labels = {
        'Tipo_Contrato': 'Tipo Contrato'
    },
    title = 'Percentual de Contratos de Clientes',
    width = 1000,
    height = 600,
    text_auto = True
)

fig.update_layout(
    font = dict(
        size = 16
    ),
    hoverlabel = dict(
        font = dict(
            size = 14
        )
    )
)

fig.update_traces(
    textfont_size = 18,
    textposition = 'outside'
)

fig.show()

    Tipo_Contrato  Contagem  Percentual
0  Month-to-month      1369       55.02
1        Two year       602       24.20
2        One year       517       20.78 



### Insight

Grande parte dos nossos clientes (55%) possuem contratos mensais conosco

### Coluna Churn

In [211]:
contagem_churn = df['Churn'].value_counts().reset_index(name = 'Contagem')
contagem_churn['Percentual'] = (contagem_churn['Contagem'] / len(df) * 100).round(2)
print(contagem_churn, '\n')

fig = px.pie(
    contagem_churn,
    values = 'Percentual',
    names = 'Churn',
    template = 'plotly',
    title = '% Clientes Churn',
    width = 1000,
    height = 600,
    hole = .5,
    category_orders = {
        'Churn': ['Yes', 'No']
    },
    color = 'Churn',
    color_discrete_map = {
        'Yes': '#FF5C5C',
        'No': '#7f7f7f'
    }
)

fig.update_traces(
    textfont = dict(
        color = 'white'
    )
)

fig.update_layout(
    hoverlabel = dict(
        font = dict(
            color = 'white',
            size = 16
        )
    ),
    font = dict(
        size = 20,
    ),
    legend = dict(
        x = 0.8
    )
)


fig.show()

  Churn  Contagem  Percentual
0    No      1841        74.0
1   Yes       647        26.0 



### Insight

Poucos clientes da nossa base são Churn, apenas 26%.

## Exercício D

In [56]:
# CALCULANDO A PORCENTAGEM DE CLIENTES CASADOS
contagem_casado = df['Casado'].value_counts().reset_index(name = 'Contagem')
contagem_casado['Percentual'] = (contagem_casado['Contagem'] / len(df) * 100).round(2)

# CALCULANDO A PORCENTAGEM DE CLIENTES COM DEPENDENTES
contagem_dependentes = df['Dependentes'].value_counts().reset_index(name = 'Contagem')
contagem_dependentes['Percentual'] = (contagem_dependentes['Contagem'] / len(df) * 100).round(2)

contagem_servico_seguranca = df['Servico_Seguranca'].value_counts().reset_index(name = 'Contagem')
contagem_servico_seguranca['Percentual'] = (contagem_servico_seguranca['Contagem'] / len(df) * 100).round(2)

print(contagem_churn, '\n')
print(contagem_idosos, '\n')
print(contagem_casado, '\n')
print(contagem_dependentes, '\n')


  Churn  Contagem  Percentual
0    No      1841        74.0
1   Yes       647        26.0 

       Idoso  Contagem  Percentual
0  Não Idoso      2086       83.84
1      Idoso       402       16.16 

  Casado  Contagem  Percentual
0     No      1262       50.72
1    Yes      1226       49.28 

  Dependentes  Contagem  Percentual
0          No      1705       68.53
1         Yes       783       31.47 



**Conclusão**: Apenas a coluna de clientes casados mostrou-se estar balanceada.

# 3) Identifique e trate as colunas que contém outliers.
Caso opte por mante-los ou altera-los justifique sua escolha.


In [188]:
#Seu código aqui
fig = px.box(
    df['Total_Pago'],
    template = 'plotly',
    width = 1000,
    height = 600,
    title = 'Boxplot de Total Pago',
)

fig.show()

In [67]:
fig = px.box(
    df['Pagamento_Mensal'],
    template = 'plotly',
    width = 1000,
    height = 600
)
fig.show()

In [69]:
fig = px.box(
    df['Tempo_como_Cliente'],
    template = 'plotly',
    width = 1000,
    height = 600
)

fig.show()

**Conclusão**: Não identifiquei nenhuma coluna com valores Outliers, portanto, não vou alterar nada nesse sentido.

# 4) Realize a etapa da análise bivariada:
A) Questione pelo menos 5 informações e traga as respostas utilizando visuais gráficos e insights.


B) Quais variáveis você acredita serem as mais importantes para esse projetos relacionadas a variável Churn?

## Exercício A

### Relação Tipo de Contrato Com Churn

In [45]:
#Seu código aqui
df_agrupado_contrato_churn = df.groupby(['Churn', 'Tipo_Contrato']).size().reset_index(name = 'Contagem')
qtd_clientes_churn = df_agrupado_contrato_churn.groupby('Churn')['Contagem'].transform('sum')

df_agrupado_contrato_churn['Percentual'] = (df_agrupado_contrato_churn['Contagem'] / qtd_clientes_churn * 100).round(2)


print(df_agrupado_contrato_churn, '\n')

fig = px.bar(
    df_agrupado_contrato_churn,
    x = 'Churn',
    y = 'Percentual',
    template = 'plotly',
    width = 1000,
    height = 600,
    color = 'Tipo_Contrato',
    title = 'Relação Churn e Tipo de Contrato'
)

fig.update_layout(
    hoverlabel = dict(
       font = dict(
            color = 'white',
            size = 16
        )
    ),
    font = dict(
        size = 18
    )
)

fig.show()

  Churn   Tipo_Contrato  Contagem  Percentual
0    No  Month-to-month       789       42.86
1    No        One year       459       24.93
2    No        Two year       593       32.21
3   Yes  Month-to-month       580       89.64
4   Yes        One year        58        8.96
5   Yes        Two year         9        1.39 



### **Insight**
Podemos observar que clientes com contrato mês a mês possuem forte propensão a serem churns. Cerca de 89% da nossa base Churn são clientes com contrato de mensal

## Relação Churn e Método de Pagamento

In [51]:
df_agrupado_churn_metodo_pgto = df.groupby(['Churn', 'Metodo_Pagamento']).size().reset_index(name = 'Contagem')
qtd_clientes_churn = df_agrupado_churn_metodo_pgto.groupby('Churn')['Contagem'].transform('sum')

df_agrupado_churn_metodo_pgto['Percentual'] = (df_agrupado_churn_metodo_pgto['Contagem'] / qtd_clientes_churn * 100).round(2)
print(df_agrupado_churn_metodo_pgto, '\n')

fig = px.bar(
    df_agrupado_churn_metodo_pgto,
    x = 'Churn',
    y = 'Percentual',
    template = 'plotly',
    color = 'Metodo_Pagamento',
    width = 1000,
    height = 600,
    title = 'Relação Churn e Método de Pagamento'
)

fig.show()

  Churn           Metodo_Pagamento  Contagem  Percentual
0    No  Bank transfer (automatic)       475       25.80
1    No    Credit card (automatic)       476       25.86
2    No           Electronic check       452       24.55
3    No               Mailed check       438       23.79
4   Yes  Bank transfer (automatic)       104       16.07
5   Yes    Credit card (automatic)        75       11.59
6   Yes           Electronic check       378       58.42
7   Yes               Mailed check        90       13.91 



### **Insight**
Cerca de 58% da base de clientes churn possuem como método de pagamento o cheque eletrônico

## Relação do Tempo Como Cliente e Churn

In [106]:
df_tempo = df.copy() # Criando uma cópia do DataFrame

# DEFININDO AS FAIXAS TEMPORAIS
valor_maximo = df_tempo['Tempo_como_Cliente'].max()
bins = list(range(0, valor_maximo + 7, 6))     # Intervalos
labels = [f'{b} a {b + 5}' for b in bins[:-1]] # Escrita das faixas

df_tempo['Faixas'] = pd.cut(
    df_tempo['Tempo_como_Cliente'],
    bins = bins,
    labels = labels,
    right = False,
    include_lowest = True
)

# CALCULANDO A QUANTIDADE DE CHURN POR FAIXA E TIRANDO O PERCENTUAL DE REPRESENTATIVIDADE
df_agrupado_tempo_churn = df_tempo.groupby(['Churn', 'Faixas']).size().reset_index(name = 'Contagem')
clientes_churn = df_agrupado_tempo_churn.groupby('Faixas')['Contagem'].transform('sum')
df_agrupado_tempo_churn['Percentual'] = (df_agrupado_tempo_churn['Contagem'] / clientes_churn * 100).round(2)

# PLOTANDO O GRÁFICO
fig = px.bar(
    df_agrupado_tempo_churn,
    x = 'Faixas',
    y = 'Percentual',
    color = 'Churn',
    color_discrete_map = {
        'Yes': '#FF5C5C',
        'No': '#7f7f7f'
    },
    template = 'plotly',
    width = 1000,
    height = 600,
    category_orders = {'Churn': ['Yes', 'No']},
    title = 'Percentual de Churn por Faixa de Tempo',
)

fig.show()

### **Insight**
Conforme mais tempo os clientes passam no nosso negócio, menor é a propensão de serem Churns

## Relação de produtos e Churn

In [4]:
df.head(5)

Genero  Idoso Casado Dependentes  Tempo_como_Cliente Servico_Internet  \
0  Female      0    Yes          No                   1              DSL   
1    Male      0     No          No                  34              DSL   
2    Male      0     No          No                   2              DSL   
3    Male      0     No          No                  45              DSL   
4  Female      0     No          No                  52               NO   

     Servico_Seguranca      Suporte_Tecnico          StreamingTV  \
0                   No                   No                   No   
1                  Yes                   No                   No   
2                  Yes                   No                   No   
3                  Yes                  Yes                   No   
4  No internet service  No internet service  No internet service   

    Tipo_Contrato           Metodo_Pagamento  Pagamento_Mensal  Total_Pago  \
0  Month-to-month           Electronic check             29.85       29.85   
1        One year               Mailed check             56.95     1889.50   
2  Month-to-month               Mailed check             53.85      108.15   
3        One year  Bank transfer (automatic)             71.45     1840.75   
4        One year               Mailed check             71.45     1022.95   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4    No

In [47]:
df_agrupado_servico_churn = df.groupby(['Churn', 'Servico_Internet']).size().reset_index(name = 'Contagem')
clientes_churn = df_agrupado_servico_churn.groupby('Churn')['Contagem'].transform('sum')
df_agrupado_servico_churn['Percentual'] = (df_agrupado_servico_churn['Contagem'] / clientes_churn * 100).round(2)

fig = px.bar(
    df_agrupado_servico_churn,
    x = 'Churn',
    y = 'Percentual',
    color = 'Servico_Internet',
    title = 'Percentual de Churn por Serviço Contratado',
    template = 'plotly',
    width = 1000,
    height = 600
)

fig.update_layout(
    hoverlabel = dict(
        font = dict(
            color = 'white',
            size = 14
        )
    ),
    font = dict(
        size = 16
    )
)

fig.show()

### **Insight**
Clientes com fibra óptica representam 72% dos clientes Churn da base, sendo um produto que provavelmente necessita de atenção

## Relação Valor Pagamento e Churn

In [117]:
df_pgto = df.copy()

# DEFININDO AS FAIXAS DE PAGAMENTO
valor_maximo = df_pgto['Pagamento_Mensal'].max()
bins = list(range(0, int(valor_maximo) + 10,11))
labels = [f'{b} a {b + 10}' for b in bins[:-1]]

df_pgto['Faixas'] = pd.cut(
    df_pgto['Pagamento_Mensal'],
    bins = bins,
    labels = labels,
    include_lowest = True,
    right = False,
)

# CALCULANDO CLIENTES CHURN POR FAIXA DE VALOR
df_agrupado = df_pgto.groupby(['Churn', 'Faixas']).size().reset_index(name = 'Contagem')
clientes = df_agrupado.groupby('Faixas')['Contagem'].transform('sum')
df_agrupado['Percentual'] = (df_agrupado['Contagem'] / clientes * 100).round(2)

fig = px.bar(
    df_agrupado,
    x = 'Faixas',
    y = 'Percentual',
    color = 'Churn',
    template = 'plotly',
    category_orders = {
        'Churn': ['Yes', 'No'],
    },
    color_discrete_map = {
        'Yes': '#FF5C5C',
        'No': '#7f7f7f'
    },
    width = 1000,
    height = 600,
)

fig.show()

### **Insight**
Clientes com faixas de mensalidade mais altas tendem a serem Churns

## Exercicio B

Depois de toda a análise feita, acredito que as variáveis:
- Servico_Internet
- Método_Pagamento
- Tipo_Contrato
- Pagamento_Mensal
- Tempo_como_Cliente